tools/PsesPsClient/LspPipe.cs

```diff
@@ -0,0 +1,564 @@
+        }
+
+        /// <summary>
+        /// Get all pending notifications.
+        /// </summary>
+        public IEnumerable<LspNotification> DrainNotifications()
+        {
+            return DrainQueue(_notificationQueue);
+        }
+
+        /// <summary>
+        /// Get all pending requests.
+        /// </summary>
+        public IEnumerable<LspRequest> DrainRequests()
+        {
+            return DrainQueue(_requestQueue);
+        }
+
+        /// <summary>
+        /// Get the next response if there is one, otherwise instantly return false.
+        /// </summary>
+        /// <param name="response">The first response in the response queue if any, otherwise null.</param>
+        /// <returns>True if there was a response to get, false otherwise.</returns>
+        public bool TryGetNextResponse(out LspResponse response)
+        {
+            return _responseBlockingOutput.TryTake(out response);
+        }
+
+        /// <summary>
+        /// Get the next response within the given timeout.
+        /// </summary>
+        /// <param name="response">The first response in the queue, if any.</param>
+        /// <param name="millisTimeout">The maximum number of milliseconds to wait for a response.</param>
+        /// <returns>True if there was a response to get, false otherwise.</returns>
+        public bool TryGetNextResponse(out LspResponse response, int millisTimeout)
+        {
+            return _responseBlockingOutput.TryTake(out response, millisTimeout);
+        }
+
+        /// <summary>
+        /// Start the pipe listener on its own thread.
+        /// </summary>
+        public void Start()
+        {
+            Task.Run(() => RunListenLoop());
+        }
+
+        /// <summary>
+        /// End the pipe listener loop.
+        /// </summary>
+        public void Stop()
+        {
+            _cancellationSource.Cancel();
+        }
+
+        /// <summary>
+        /// Stops and disposes the pipe listener.
+        /// </summary>
+        public void Dispose()
+        {
+            Stop();
+            _stream.Dispose();
+        }
+
+        private async Task RunListenLoop()
+        {
+            CancellationToken cancellationToken = _cancellationSource.Token;
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                LspMessage msg = await ReadMessage().ConfigureAwait(false);
+                switch (msg)
+                {
+                    case LspNotification notification:
+                        _notificationQueue.Enqueue(notification);
+                        continue;
+
+                    case LspResponse response:
+                        _responseBlockingOutput.Add(response);
+                        continue;
+
+                    case LspRequest request:
+                        _requestQueue.Enqueue(request);
+                        continue;
+                }
+            }
+        }
+
+        private async Task<LspMessage> ReadMessage()
```

Comment:
How much of this is borrowed from PSES proper? Might be worth calling out what is

Reply:
To get an MVP I basically stole the message formation code, which wasn't very reusable from the DLL because it has a lot of logging and some other stuff in it.

The message send/receive code I rewrote, because PSES does some of it strangely.

I'm not a huge fan of the way PSES does a lot of it, especially the types it uses for messages. So I'm hoping the borrowing is a short/medium-term thing

Reply:
Do you think the stuff you rewrote should be pulled into PSES proper since it does strange things?

Reply:
Yeah that's what I'm gearing up for basically; finding out what works and what doesn't to see what could be done better withing PSES itself